app/models/private_message_topic_tracking_state.rb

```diff
@@ -0,0 +1,209 @@
+# frozen_string_literal: true
+
+# This class is used to mirror unread and new status for private messages between
+# server and client.
+#
+# On the server side, this class has two main responsibilities. The first is to
+# query the database for the initial state of a user's unread and new private
+# messages. The second is to publish message_bus messages to notify the client
+# of various topic events.
+#
+# On the client side, we have a `PrivateMessageTopicTrackingState` class as well
+# which will load the initial state into memory and subscribes to the relevant
+# message_bus messages. When a message is received, it modifies the in-memory
+# state based on the message type. The filtering for new and unread topics is
+# done on the client side based on the in-memory state in order to derive the
+# count of new and unread topics efficiently.
+class PrivateMessageTopicTrackingState
+  CHANNEL_PREFIX = "/private-message-topic-tracking-state"
+  NEW_MESSAGE_TYPE = "new_topic"
+  UNREAD_MESSAGE_TYPE = "unread"
+  ARCHIVE_MESSAGE_TYPE = "archive"
+  GROUP_ARCHIVE_MESSAGE_TYPE = "group_archive"
+
+  def self.report(user)
+    sql = new_and_unread_sql(user)
+
+    DB.query(
+      sql + "\n\n LIMIT :max_topics",
+      {
+        max_topics: TopicTrackingState::MAX_TOPICS,
+        min_date: Time.at(SiteSetting.min_new_topics_time).to_datetime
+      }
+    )
+  end
+
+  def self.new_and_unread_sql(user)
+    sql = report_raw_sql(user, skip_unread: true)
+    sql << "\nUNION ALL\n\n"
+    sql << report_raw_sql(user, skip_new: true)
+  end
+
+  def self.report_raw_sql(user, skip_unread: false,
+                                skip_new: false)
+
+    unread =
+      if skip_unread
+        "1=0"
+      else
+        TopicQuery.unread_filter(Topic, staff: user.staff?).where_clause.ast.to_sql
+      end
+
+    new =
+      if skip_new
+        "1=0"
+      else
+        new_filter_sql
+      end
+
+    sql = +<<~SQL
```

Comment:
I do worry a little bit about drift between the logic and queries in the two tracking classes, but I doubt we will be adding more so I think the duplication works well for now. Maybe add a note to the top of each class with the other comments to say "When changing queries or logic here, be sure to check OTHER_TRACKING_CLASS at the same time"

Reply:
Good point. I realized there is a fair bit of duplication between the two classes so I cleaned that up such that `PrivateMessageTopicTrackingState` will rely on `TopicTrackingState` for the new and unread SQL queries when it constructs its SQL string.